docs/proposals/PROPOSAL_E.md

```diff
@@ -0,0 +1,296 @@
+
+```jsonc
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "config": { ... },
+  "layers": [ ... ],
+  "reference": { // new field
+    "mediaType": "application/vnd.oci.image.manifest.v1+json", // any manifest media type
+    "size": 1234,
+    "digest": "sha256:a1a1a1...",
+    "annotations": []
+  },
+  "annotations": [
+    // annotations for this artifact here
+  ]
+}
+```
+
+#### Artifact Spec
+
+Create a new artifact media type to support future use cases where a separate config blob, and an ordered list of blobs, may not match an artifact's use case:
+
+```jsonc
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.artifact.manifest.v1+json",
+  "blobs": [
+    // optional list, ordering is not enforced by the spec
+    // but may be required for specific artifact types
+  ],
+  "reference": { // optional
+    // include reference here
+  },
+  "annotations": [ // optional
+    // annotations for this artifact here
+  ]
+}
+```
+
+### Registry HTTP API
+
+#### References API
+
+Add the following `references` API (`<ref>` is the digest in the `reference` schema field):
+
+```text
+GET /v2/<name>/references/<ref>
+```
+
+The response is an Index of descriptors:
+
+```jsonc
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.index.v1+json",
+  "manifests": [
+    {
+      "mediaType": "application/vnd.oci.image.manifest.v1+json",
+      "size": 1234,
+      "digest": "sha256:a1a1a1...",
+      "annotations": [
+        // annotations pulled up from manifest
+        "org.opencontainers.artifact.type": "example/icecream",
+        "org.opencontainers.artifact.created": "2022-01-01T14:42:55Z",
+        "org.example.icecream.flavor": "chocolate"
+      ]
+    },
+    {
+      "mediaType": "application/vnd.oci.image.manifest.v1+json",
+      "size": 1234,
+      "digest": "sha256:a2a2a2...",
+      "annotations": [
+        "org.opencontainers.artifact.type": "example/icecream",
+        "org.opencontainers.artifact.created": "2022-01-01T15:24:30Z",
+        "org.example.icecream.flavor": "vanilla"
+      ]
+    }
+  ],
+  "annotations": [
+    "org.opencontainers.references.params": "..." // optional base64 encoded parameters
+  ]
+}
+```
+
+If a registry does not implement the `references` API, it MUST return a 404.
+If a query results in no references found, an empty manifest list MUST be returned.
+
+#### Ordering and Pagination
+
+- The registry MUST order the results in a way that allows pagination.
+- Adding the `n` query parameter is used to limit the number of entries per index returned.
+- The registry SHOULD return fewer than `n` results when the generated Index would exceed the recommended maximum manifest size.
+- The `Link` HTTP header is included in the response when additional results are available and is set to the URL for the next page of results.
+
+#### Filtering and Sorting
```

Comment:
This is the main part I'm getting tripped up on. I think that allowing filtering and sorting on any attribute ends up complicating the implementation of this unnecessarily. I also think that having this as a SHOULD will make for a varied and unpredictable experience for clients when moving between registries and probably cause bugs due to difficulties in testing. In other proposals we looked at explicitly requiring filtering or ordering on specific attributes only. Could we specify a subset of annotations or fields that registries should be prepared to index on and clients can expect to work?

Reply:
Individual registries could limit themselves to only a subset of annotations, and a client seeing that not all annotations were filtered or sorted would then been required to handle the extra processing themselves. Clients validate the filters/sorting provided in the `org.opencontainers.references.params` annotation, so going between registries should be handled correctly. It won't be two code paths for clients, but an optimization to skip a path when a registry indicates it's already done that work.

Reply:
This synthetic index does not fit well with the registry data model, it has no stable URL, is not versioned, no stable hash so no content addressability, and cannot be referred to by another registry object.

Reply:
I think the consensus yesterday (Brandon was out for it) was to drop filtering from this proposal for now.

Reply:
Should consider the schema that proposal A has  - https://github.com/opencontainers/wg-reference-types/blob/main/docs/proposals/PROPOSAL_A.md#api-path @sudo-bmitch